common/src/main/java/com/microsoft/identity/common/internal/ui/webview/switchbrowser/SwitchBrowserUriHelper.kt

```diff
@@ -69,51 +79,53 @@ object SwitchBrowserUriHelper {
+        if (STATE_VALIDATION_REQUIRED) {
+            val state = uri.getQueryParameter(
+                SWITCH_BROWSER.STATE
+            )
+            if (state.isNullOrEmpty()) {
+                // This should never happen, but if it does, we should log it and throw.
+                val errorMessage = "switch browser action state is null or empty"
+                val exception = ClientException(ClientException.MALFORMED_URL, errorMessage)
+                Logger.error(methodTag, errorMessage, exception)
+                throw exception
+            } else {
+                queryParams[SWITCH_BROWSER.STATE] = state
+            }
+        }
         // Construct the uri to the process endpoint.
         return buildSwitchBrowserUri(actionUri, queryParams)
     }
 
-    /**
-     * Build a generic switch browser uri.
-     *
-     * @param actionUri The action uri to be opened.
-     * @param queryParams The query parameters to be included in the switch browser uri.
-     *
-     * @return The switch browser uri constructed from the action uri and query parameters.
-     */
-    @Throws(IllegalArgumentException::class, NullPointerException::class, UnsupportedOperationException::class)
-    private fun buildSwitchBrowserUri(
-        actionUri: String,
-        queryParams: HashMap<String, String> = hashMapOf()
-    ): Uri {
-        val paths = actionUri.split("/")
-        val authority = paths[0]
-        val uriBuilder = Uri.Builder()
-            .scheme("https")
-            .encodedAuthority(authority)
-        for (i in 1 until paths.size) {
-            uriBuilder.appendPath(paths[i])
-        }
-        for ((key, value) in queryParams.entries) {
-            uriBuilder.appendQueryParameter(key, value)
-        }
-        return uriBuilder.build()
-    }
-
     /**
      * Build the resume uri for the switch browser challenge.
      *
      * @param actionUri The action uri to be opened.
+     * @param state The state to be included in the switch browser uri.
      *
      * @return The resume uri constructed from the bundle.
      * e.g. actionUri
      */
-    fun buildResumeUri(actionUri: String): Uri {
+    fun buildResumeUri(actionUri: String, state: String?): Uri {
+        val methodTag = "$TAG:buildResumeUri"
         // Construct the uri to the resume endpoint.
-        return buildSwitchBrowserUri(actionUri)
+        val queryParams = hashMapOf<String, String>()
+        if (STATE_VALIDATION_REQUIRED) {
+            if (state.isNullOrEmpty()) {
+                // This should never happen, but if it does, we should log it and throw.
+                val errorMessage = "State is null or empty"
+                val exception = ClientException(ClientException.MISSING_PARAMETER, errorMessage)
+                Logger.error(methodTag, errorMessage, exception)
+                throw exception
+            } else {
+                queryParams[SWITCH_BROWSER.STATE] = state
+            }
```

Comment:
I see similar code above, perhaps it can be refactored into its own method

Reply:
[done](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/2631/commits/f84da49bc1a22a471ef29286d0acc8643a079853)